### PR TITLE
[SPARK-8996] [MLlib] [PySpark] Python API for Kolmogorov-Smirnov Test

### PR DESCRIPTION
**Pull request by MechCoder**
Python API for the KS-test

Statistics.kolmogorovSmirnovTest(data, distName, *params)
I'm not quite sure how to support the callable function since it is not serializable.

### CODE REVIEW

**Comment by mengxr** [comment]
@sryza @josepablocam Could you help review this PR? Thanks!

**Comment by josepablocam** [comment]
@MechCoder I'm not as familiar with Python, so this might be a horrid suggestion, but one option would be to require users to pass in a named function, rather than a lambda. cPickle can handle named functions (but not lambdas). I suspect there is a better suggestion out there for this though...

```
>>> import pickle
>>> from scipy.stats.distributions import norm
>>> def mynorm(x):
...   norm.cdf(x)
... 
>>> pickle.dumps(mynorm)
'c__main__\nmynorm\np0\n.'
```

**Comment by MechCoder** [comment]
Hmm. I did try that out, but there was some problem with python to java conversion. I'll have a deeper look.

**Comment by MechCoder** [comment]
It would also be really helpful if you could give a quick pass over the rest of the PR

File: python/pyspark/mllib/stat/_statistics.py [hunk 1]
```diff
@@ -238,6 +242,54 @@ def chiSqTest(observed, expected=None):
...
+        is calculated which is the number of points having a CDF value lesser than a given point
+        divided by the total number of points. Since the data is sorted, this is a step function
+        that rises by (1 / length of data) for every ordered point.
+
+        The KS statistic gives us the maximum distance between the ECDF and the CDF. Intuitively
+        if this value is large, the probabilty that the null hypothesis is true becomes small.
+        For specific details of the implementation, please have a look at the Scala documentation.
+
+        :param data: RDD, samples from the data
+        :param distName: string, currently only "norm" is suuported. (Normal distribution)
```
Comment (josepablocam) [inline python/pyspark/mllib/stat/_statistics.py]: suuported -> supported

File: mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala [hunk 2]
```diff
@@ -1093,6 +1093,27 @@ private[python] class PythonMLLibAPI extends Serializable {
...
+   * Wrapper around Statistics.kolmogorovSmirnovTestWrapper.
+   */
+  def kolmogorovSmirnovTestWrapper(
+      data: JavaRDD[Double],
+      distName: String,
+      params: JList[Double]): KolmogorovSmirnovTestResult = {
+    val seqParams = params.asScala.toSeq
+    Statistics.kolmogorovSmirnovTest(data, distName, seqParams: _*)
+  }
+
```
Comment (josepablocam) [inline mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala]: I'll defer to @mengxr and @sryza but just wondering: do we need the 2 separate wrappers? seems to me that 1 should do (the last one). If params is passed in empty on the python side, that should just be an empty list, and the asScala.toSeq, will make it an empty sequence.

Comment (MechCoder) [inline mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala]: It seems like it. I'll make the change.

**Comment by mengxr** [comment]
@josepablocam @MechCoder It should be sufficient to support only the `distName` + `params` API in this PR.

File: python/pyspark/mllib/stat/_statistics.py [hunk 3]
```diff
@@ -238,6 +242,54 @@ def chiSqTest(observed, expected=None):
...
+    @ignore_unicode_prefix
+    def kolmogorovSmirnovTest(data, distName="norm", *params):
+        """
+        .. note:: Experimental
+
+        Performs the Kolmogorov Smirnov (KS) test for data sampled from a continuous
+        distribution. It tests the null hypothesis that the data is generated from a
+        particular distribution.
+
+        The given data is sorted, the Empirical Cumulative Distribution Function (ECDF)
```
Comment (mengxr) [inline python/pyspark/mllib/stat/_statistics.py]: pydoc should have max line width 72 per PEP8.

File: mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala (hunk 2 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala]: +1 on only one wrapper. And we don't need to call the method `Wrapper`. Just follow how `chiSqTest` is implemented.

**Comment by MechCoder** [comment]
I've addressed your comments. please review

File: python/pyspark/mllib/stat/_statistics.py [hunk 4]
```diff
@@ -238,6 +242,60 @@ def chiSqTest(observed, expected=None):
             jmodel = callMLlibFunc("chiSqTest", _convert_to_vector(observed), expected)
         return ChiSqTestResult(jmodel)
 
+    @staticmethod
+    @ignore_unicode_prefix
+    def kolmogorovSmirnovTest(data, distName="norm", *params):
+        """
+        .. note:: Experimental
+
+        Performs the Kolmogorov Smirnov (KS) test for data sampled from
```
Comment (mengxr) [inline python/pyspark/mllib/stat/_statistics.py]: `Kolmogorov Smirnov` -> `Kolmogorov-Smirnov`

**Comment by mengxr** [comment]
LGTM except one minor comment.

File: python/pyspark/mllib/stat/_statistics.py [hunk 5]
```diff
@@ -238,6 +242,60 @@ def chiSqTest(observed, expected=None):
...
+                         (Normal distribution) to calculate the
+                         theoretical distribution of the data.
+        :param params: additional values which need to be provided for
+                       a certain distribution.
+                       If not provided, the default values are used.
+        :return: KolmogorovSmirnovTestResult object containing the test
+                 statistic, degrees of freedom, p-value,
+                 the method used, and the null hypothesis.
+
+        >>> kstest = Statistics.kolmogorovSmirnovTest
```
Comment (sryza) [inline python/pyspark/mllib/stat/_statistics.py]: Small thing: can we include an example that passes parameters in?

**Comment by sryza** [comment]
LGTM modulo a small comment

**Comment by MechCoder** [comment]
done

**Comment by mengxr** [comment]
Merged into master. Thanks!

**Pull request closed without merging**